### DISCUSSION TITLE
add authentication step for swagger before open the the page.

### DISCUSSION THREAD

**Discussion by wangxin688** [fastapi/fastapi#9230] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
no code implemented
```

### Description

when we deploy the server in production, we want to add an limitation for swagger UI access, hide to enduser without granted access. Currently, the `docs` or 'redocs' will expose to users directly if user already know the path of api docs. So, is there any chance to add pre-authentication with simple verification? or How can I implement with this feature?

### Operating System

macOS

### FastAPI Version

0.92.0

### Python Version

3.9.6

**Comment by yinziyan1206** [comment]
I think it's better to use nginx to set `baseauth` for openapi pages

**Reply by wangxin688** [reply, marked as answer]
I just figured it out. Same logic as redoc and openapi.json
```python
from fastapi import Depends, FastAPI, HTTPException
from fastapi.security import HTTPBasic, HTTPBasicCredentials
from starlette.status import HTTP_401_UNAUTHORIZED
from fastapi.openapi.docs import get_swagger_ui_html
app = FastAPI(
    docs_url=None
)

security = HTTPBasic()

@app.get("/docs",include_in_schema=False)
async def get_documentation(credentials: HTTPBasicCredentials = Depends(security)):
    if credentials.username != "someone" or credentials.password != "password":
        raise HTTPException(
            status_code=HTTP_401_UNAUTHORIZED,
            detail="Incorrect email or password",
            headers={"WWW-Authenticate": "Basic"},
        )
    else:
        return get_swagger_ui_html(openapi_url="/openapi.json", title="docs")

**Reply by allanzolnoi** [reply]
hey @wangxin688 this worked for me on local, but when deployed to our dev environment it didnt show the auth popup, any idea as to what could be the issue?

**Reply by bikmetle** [reply]
In my case, the issue was in the exception_handlers in FastAPI. The problem was that the exception handler created a new JSONResponse without preserving the headers.

**Comment by syedamanat** [comment]
If any of you are still seeking an answer to add a popup authentication before accessing swagger `/docs`; you can follow the below 

`sudo apt-get install apache2-utils`
`sudo htpasswd -c /etc/nginx/.htpasswd username`
username would be what you enter in `username` field of the popup and password respectively 

Modify your nginx like below 

```
location /swagger-ui {
        auth_basic "Restricted Content";
        auth_basic_user_file /etc/nginx/.htpasswd;  # Path to the password file
        alias /path/to/swagger-ui/;  # Ensure this path points to your Swagger UI directory; configuration does not change incase of a proxy_pass
        try_files $uri $uri/ =404;
    }
```

`sudo nginx -t` to verify nginx configuration
`sudo systemctl reload nginx` 

Hope it helps! Cheers.
